pkg/cache/scheduler_callback_test.go

```diff
@@ -0,0 +1,653 @@
+		return app.sm.Current() == ApplicationStates().Resuming
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Resuming state")
+}
+
+func TestUpdateApplication_Resuming_AppNotReserving(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Submitted) // not in Reserving - nothing should happen
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: appID,
+				State:         ApplicationStates().Resuming,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+}
+
+func TestUpdateApplication_Resuming_AppNotFound(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Reserving)
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: "nonExisting",
+				State:         ApplicationStates().Resuming,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+}
+
+func TestUpdateApplication_Failing(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Running)
+	NewPlaceholderManager(context.apiProvider.GetAPIs())
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: appID,
+				State:         ApplicationStates().Failing,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	err = utils.WaitForCondition(func() bool {
+		return app.sm.Current() == ApplicationStates().Failing
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Failing state")
+}
+
+func TestUpdateApplication_Failed(t *testing.T) {
+	callback, context := initCallbackTest(t, false, false)
+	defer dispatcher.UnregisterAllEventHandlers()
+	defer dispatcher.Stop()
+	app := context.getApplication(appID)
+	app.sm.SetState(ApplicationStates().Running)
+	NewPlaceholderManager(context.apiProvider.GetAPIs())
+
+	err := callback.UpdateApplication(&si.ApplicationResponse{
+		Updated: []*si.UpdatedApplication{
+			{
+				ApplicationID: appID,
+				State:         ApplicationStates().Failed,
+			},
+		},
+	})
+	assert.NilError(t, err, "error updating application")
+	err = utils.WaitForCondition(func() bool {
+		return app.sm.Current() == ApplicationStates().Failing
+	}, 10*time.Millisecond, time.Second)
+	assert.NilError(t, err, "application has not transitioned to Failing state")
+}
+
+func TestUpdateNode_Accepted(t *testing.T) {
```

Comment:
It seems `TestUpdateNode_Accepted` is similar to `TestUpdateNode_Rejected`. Maybe we can merge them into single test case?

Reply:
Done.